### PR TITLE
Training engine: implement eval_step API

### PR DESCRIPTION
**Pull request by SurbhiJainUSC**
# Description

* Implements `eval_step()` API in `MaxTextTrainingEngine` mirroring tunix's peft_trainer_v2. 
* MetricsRecorder takes a `mode` argument and the engine owns a second mode="eval" instance, keeping eval numbers out of the train buffer and out of the checkpoint; MetricsBuffer.mode then flows through write_metrics into both the console line and the TensorBoard tag, which becomes {mode}/{name}. 
* eval_context() accumulates across every micro-batch into one buffer (WeightedMetric appends sum and denominator, so compute() is the weighted mean over the whole eval set) and writes it exactly once on exit.

# Tests

CI tests

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by khatwanimohit — APPROVED** [review]

File: src/maxtext/training_engine/maxtext_engine.py [hunk 1]
```diff
@@ -1142,29 +1207,59 @@ def update(self, **kwargs: Any) -> int:
...
+    if self._compile_requested:
+      dynamic_batch, static_batch = _split_static_and_dynamic(batch)
+      signature = _batch_signature(dynamic_batch, static_batch)
+      if self._compiled_eval is None or self._needs_recompile(signature, self._compiled_eval_signature):
+        self._compile_eval_for_batch(dynamic_batch, static_batch)
+      loss, aux = self._compiled_eval(params, rest, dynamic_batch)
+    else:
+      loss, aux = self._eval_kernel(params, rest, batch)
+
+    if isinstance(loss, abstract_engine.WeightedMetric):
```
Comment (igorts-git) [inline src/maxtext/training_engine/maxtext_engine.py:1265]: Are there situations where loss won't be WeightedMetric? Should we support those cases as well? Otherwise we are not printing anything.

File: src/maxtext/training_engine/maxtext_engine.py [hunk 2]
```diff
@@ -1142,29 +1207,59 @@ def update(self, **kwargs: Any) -> int:
...
+      self.record_metrics("loss", loss, mode=metrics_module.Mode.EVAL)
```
Comment (igorts-git) [inline src/maxtext/training_engine/maxtext_engine.py:1266]: Remind me, do this log immediately (and blocks on the JAX to finish execution) or only adds the loss to the buffer?

File: src/maxtext/training_engine/maxtext_engine.py [hunk 3]
```diff
@@ -1142,29 +1207,59 @@ def update(self, **kwargs: Any) -> int:
...
     """
-    if not self._eval_step_warned:
-      self._eval_step_warned = True
-      logging.warning(
-          "MaxTextTrainingEngine.eval_step is not implemented: it evaluates nothing and "
-          "records no metrics, so any eval result reported for this run is meaningless. "
-          "Logged once per engine instance."
-      )
+    batch = self._prepare_batch(payload)
+
```
Comment (igorts-git) [inline src/maxtext/training_engine/maxtext_engine.py:1242]: Should we use `self._throttler.wait_for_all()` here and `self._throttler.add_computation()` at the end of this function?

**Review by igorts-git — APPROVED** [review]

File: src/maxtext/training_engine/maxtext_engine.py (hunk 2 above)
Comment (SurbhiJainUSC) [reply src/maxtext/training_engine/maxtext_engine.py:1266]: It only adds the compuation to the buffer, wihtout blocking.

File: src/maxtext/training_engine/maxtext_engine.py (hunk 1 above)
Comment (SurbhiJainUSC) [reply src/maxtext/training_engine/maxtext_engine.py:1265]: I think we initially added this to support MaxText default loss function with this training engine, where the loss value is not an object of `WeightedMetric`.

File: src/maxtext/training_engine/maxtext_engine.py (hunk 3 above)
Comment (SurbhiJainUSC) [reply src/maxtext/training_engine/maxtext_engine.py:1242]: Good catch, added rate-limiting in eval to prevent Host OOMs on large eval sets.

**Review by igorts-git — APPROVED** [review]
